consensus/src/zk.rs

```diff
@@ -0,0 +1,179 @@
+use std::process;
+use std::time::Duration;
+use std::thread::{self, Thread};
+
+use serde::Serialize;
+use serde::de::DeserializeOwned;
+use serde_json;
+use zookeeper::{Acl, CreateMode, KeeperState, WatchedEvent, Watcher, ZkError, ZooKeeper};
+
+use Authority;
+use Epoch;
+use CONTROLLER_KEY;
+
+struct EventWatcher;
+impl Watcher for EventWatcher {
+    fn handle(&self, e: WatchedEvent) {
+        if e.keeper_state != KeeperState::SyncConnected {
+            eprintln!("Lost connection to ZooKeeper! Aborting");
+            process::abort();
+        }
+    }
+}
+
+/// Watcher which unparks the thread that created it upon triggering.
+struct UnparkWatcher(Thread);
+impl UnparkWatcher {
+    pub fn new() -> Self {
+        UnparkWatcher(thread::current())
+    }
+}
+impl Watcher for UnparkWatcher {
+    fn handle(&self, _: WatchedEvent) {
+        self.0.unpark();
+    }
+}
+
+pub struct ZookeeperAuthority {
+    zk: ZooKeeper,
+}
+
+impl ZookeeperAuthority {
+    /// Create a new instance.
+    pub fn new(connect_string: &str) -> Self {
+        let zk = ZooKeeper::connect(connect_string, Duration::from_secs(1), EventWatcher).unwrap();
+        let _ = zk.create(
+            "/",
+            vec![],
+            Acl::open_unsafe().clone(),
+            CreateMode::Persistent,
+        );
+        Self { zk }
+    }
+}
+
+impl Authority for ZookeeperAuthority {
+    fn become_leader(&self, payload_data: Vec<u8>) -> Epoch {
+        loop {
+            if let Some(epoch) = self.zk
+                .create(
```

Comment:
Checking my understanding: we currently only support a single Soup deployment running against the same Zookeeper, i.e., there are no unique identifiers for different deployments. Correct?

I ask partly because it'd be nice to support multiple deployments, and partly because one could imagine that leftover Zookeeper state from a previous run is still around at this key as the lease hasn't expired yet, but we've already restarted Soup. I suspect the latter isn't a problem as we include a nonce in the payload (IIRC), so we won't be misled into believing that the dead leader is the current controller. Correct?